### PR TITLE
Synchronize accounting for render op depths

### PR DESCRIPTION
**Pull request by flar**
Experimental Canvas was getting depth assertion errors while trying to use the depth values supplied by DisplayList. This was mainly due to a difference in understanding as to how many depth values to allocate to a drawImageNine operation. 

n case there are additional discrepancies, debugging code is added to assert the understanding of how many depth values the experimental canvas uses on each rendering operation. The depth debugging can be turned on and off with a #define

### CODE REVIEW

File: impeller/display_list/dl_dispatcher.cc [hunk 1]
```diff
@@ -36,6 +36,53 @@
...
+  const std::string file_;
+  const int line_;
+
+  const impeller::Canvas& canvas_;
+  const uint64_t allowed_;
+  const uint64_t old_depth_;
+  const uint64_t old_max_;
+};
+
+#define AUTO_DEPTH_WATCHER(d) \
```
Comment (jonahwilliams) [inline impeller/display_list/dl_dispatcher.cc:75]: Can you document, here or elsewhere how to use AUTO_DEPTH_WATCHER if I were to add a new canvas command. it just takes the required depth increment, right?

File: display_list/dl_op_records.h [hunk 2]
```diff
@@ -790,6 +790,7 @@ struct DrawImageRectOp final : DrawOpBase {
 #define DEFINE_DRAW_IMAGE_NINE_OP(name, render_with_attributes)            \
   struct name##Op final : DrawOpBase {                                     \
     static constexpr auto kType = DisplayListOpType::k##name;              \
+    static constexpr uint32_t kDepthInc = 9;                               \
```
Comment (jonahwilliams) [inline display_list/dl_op_records.h:793]: Does this get used somewhere?

Comment (flar) [reply display_list/dl_op_records.h:793]: In `DisplayListBuilder::Push`

File: impeller/display_list/dl_dispatcher.cc (hunk 1 above)
Comment (flar) [reply impeller/display_list/dl_dispatcher.cc]: Done

**Comment by jonahwilliams** [comment]
Looks go to me but not to CI. Also we should turn off the depth logging for profile/release, right?

**Comment by flar** [comment]
It looks like the test mechanism caught a number of bugs that are still in the system.

I was going to ask about the disposition of that mechanism moving forward. The asserts in the ExpCanvas are enabled for all builds. These are similar, but are far more numerous. I could make it debug only (once I fix these failures).

**Comment by jonahwilliams** [comment]
I think it would be good to shake out the bugs now. Longer term we should probably not FML_CHECK unless we think something will cause crashes elsewhere.

**Comment by flar** [comment]
> I think it would be good to shake out the bugs now. Longer term we should probably not FML_CHECK unless we think something will cause crashes elsewhere.

I just realized that these tests are running without ExpCanvas, which means that the assertions are trying to apply to the entity system that consumes depth values differently. I think there is at least one case where the entity canvas pushes an extra entity - which doesn't matter because it isn't using the depth accounting from DL, it is using its own.

So, I'm going to push a couple of commits here to test things out:

- One with the mechanism enabled but the exp canvas also enabled (does this have any hope of passing?) to see if it still triggers the assertions in the new mechanism - this step will be just "for our information"
- Another with the mechanism "enabled", but the exp canvas disabled and the mechanism gated behind requiring both the mechanism enable flag and also the exp canvas enable flag - which hopefully will not have errors because the mechanism won't be used with the non-exp canvas
- Finally, a commit with the mechanism gated by debug mode and only enabled when using exp canvas.

**Comment by jonahwilliams** [comment]
> One with the mechanism enabled but the exp canvas also enabled (does this have any hope of passing?) to see if it still triggers the assertions in the new mechanism - this step will be just "for our information"

One would hope, as this is what I'd like to opt into after this and https://github.com/flutter/engine/pull/54604 land this week.

**Comment by flar** [comment]
I tried running the tests locally and they took a really long time and then died anyway due to vulkan issues on my machine so I just jumped to the end state and pushed that - hopefully it will pass CI tests this time (and continue to do so when exp canvas is enabled globally).

**Comment by flar** [comment]
Passing CI now, depth watcher now restricted to ExpCanvas + DEBUG

**Review by jonahwilliams — APPROVED** [review]
LGTM!

**Pull request merged**